test/extensions/filters/http/wasm/wasm_filter_integration_test.cc

```diff
@@ -164,8 +190,34 @@ TEST_P(WasmFilterIntegrationTest, BodyManipulation) {
 
   Http::TestResponseHeaderMapImpl expected_response_headers{{":status", "200"}};
 
-  runTest(request_headers, "request_body", expected_request_headers, "replace",
-          upstream_response_headers, "response_body", expected_response_headers, "replace");
+  auto request_body = std::vector<std::string>{{"request_body"}};
+  auto upstream_response_body = std::vector<std::string>{{"upstream_body"}};
+  runTest(request_headers, request_body, expected_request_headers, "replace",
+          upstream_response_headers, upstream_response_body, expected_response_headers, "replace");
+}
+
+TEST_P(WasmFilterIntegrationTest, BodyBufferedManipulation) {
+  setupWasmFilter("", "body");
+  HttpIntegrationTest::initialize();
+
+  Http::TestRequestHeaderMapImpl request_headers{{":scheme", "http"},
+                                                 {":method", "GET"},
+                                                 {":path", "/"},
+                                                 {":authority", "host"},
+                                                 {"x-test-operation", "SetEndOfBodies"}};
+
+  Http::TestRequestHeaderMapImpl expected_request_headers{{":path", "/"}};
+
+  Http::TestResponseHeaderMapImpl upstream_response_headers{{":status", "200"},
+                                                            {"x-test-operation", "SetEndOfBodies"}};
+
+  Http::TestResponseHeaderMapImpl expected_response_headers{{":status", "200"}};
+
+  auto request_body = std::vector<std::string>{{"request_"}, {"body"}};
+  auto upstream_response_body = std::vector<std::string>{{"upstream_"}, {"body"}};
```

Comment:
could we add a test with 3 (or more) pieces?

For the first piece, the addDecodedData won't be called before it haven't enter the buffering mode.
For the second piece, the addDecodeData will be called but it's not the ending, so, the filter chain will still be stoped to handle the empty piece (the piece has been moved by the addDecodedData).
For the third piece, the addDecodedData will be called and it the ending, so the filter chain will continue.